Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
+
+    CartesianPower(InputIt first_, InputIt last_, size_t power_);
+    template<typename Container>
+    CartesianPower(const Container& container, size_t power_);
```

Comment:
OK, I guess "all constructor parameters have a trailing `_` and all member names do not" is a simpler style guideline.

`power` could be made private; a warning should be emitted when trying to set it to a value other than 3, 4, or 6.

It's not really possible to validate the iterators dynamically. If `size` or `base_size` were made a data member of `CartesianProduct`, the iterators should be made private to enforce updates to `size`. Currently that's not the case.